### PR TITLE
Porting the GMM tutorial

### PR DESCRIPTION
**Pull request by ordabayevy**
Porting [the GMM tutorial](https://pyro.ai/examples/gmm.html) from Pyro. I was able to reproduce everything except tracking gradient norms with `.register_hook`. I had to adjust `infer_discrete` so it would work with models that have non-enumerated variables (global variables in this case).

### CODE REVIEW

**Comment by ordabayevy** [comment]
Hmm ... failing tests (`test/infer/test_mcmc.py::test_binomial_stable_x64` and `test/infer/test_mcmc.py::test_beta_bernoulli_x64[SA]`) pass locally on my computer.

**Comment by fehiepsi** [comment]
@ordabayevy You can use the follow hook to collect gradient norms
```python
from collections import defaultdict
import optax
import jax
import jax.numpy as jnp

def hook_optax(optimizer):
    gradient_norms = defaultdict(list)

    def append_grad(grad):
        for name, g in grad.items():
            gradient_norms[name].append(float(jnp.linalg.norm(g)))
        return grad

    def update_fn(grads, state, params=None):
        grads = jax.pure_callback(append_grad, grads, grads)
        return optimizer.update(grads, state, params=params)

    return optax.GradientTransformation(optimizer.init, update_fn), gradient_norms

optimizer, gradient_norms = hook_optax(optax.adam(0.1))
params = {"x": 0.}
opt_state = optimizer.init(params)

def loss(params):
    return (params["x"] - 2.) ** 2

@jax.jit
def step(params, opt_state):
    loss_value, grads = jax.value_and_grad(loss)(params)
    updates, opt_state = optimizer.update(grads, opt_state, params)
    params = optax.apply_updates(params, updates)
    return params, opt_state, loss_value

for i in range(100):
    params, opt_state, loss_value = step(params, opt_state)
params
```

**Comment by ordabayevy** [comment]
Thanks @fehiepsi! Updated the tutorial to include gradient norms. Here are the plots I get:
norms1
norms2

 The first plot deviates a bit from the Pyro tutorial one, but maybe it is okay?

**Review by fehiepsi — APPROVED** [review]
Awesome work! Thanks, Yerdos!

File: notebooks/source/gmm.ipynb [hunk 1]
```diff
@@ -0,0 +1,853 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Text(0.5, 1.0, 'Convergence of SVI')"
```
Comment (fehiepsi) [inline notebooks/source/gmm.ipynb:246]: You can add `plt.show()` at the end of each cell block to remove those lines.

File: notebooks/source/gmm.ipynb [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+    "\n",
+    "    def update_fn(grads, state, params=None):\n",
+    "        grads = pure_callback(append_grad, grads, grads)\n",
+    "        return optimizer.update(grads, state, params=params)\n",
+    "\n",
+    "    return optax.GradientTransformation(optimizer.init, update_fn), gradient_norms\n",
+    "\n",
+    "\n",
+    "optim, gradient_norms = hook_optax(optax.adam(learning_rate=0.1, b1=0.8, b2=0.99))\n",
+    "global_svi = SVI(model, global_guide, optax_to_numpyro(optim), loss=elbo)"
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply notebooks/source/gmm.ipynb]: Didn't know that!

File: numpyro/contrib/funsor/discrete.py [hunk 3]
```diff
@@ -71,19 +71,7 @@ def _sample_posterior(
...
-            output = funsor.Reals[node["fn"].event_shape]
-            value = funsor.to_funsor(
-                node["value"], output, dim_to_name=node["infer"]["dim_to_name"]
-            )
-            value = value(**sample_subs)
-            node["value"] = funsor.to_data(
-                value, name_to_dim=node["infer"]["name_to_dim"]
-            )
-        else:
+        if node["infer"].get("enumerate") == "parallel":
```
[the remark this answers is not in the document]
Comment (ordabayevy) [reply numpyro/contrib/funsor/discrete.py:74]: There are non-enumerated latent variables in the GMM tutorial that trigger the `else` branch here but since these variables are not enumerated `_get_support_value(log_measure, name)` will give an error.

I think we only need to get support values for enumerated variables. For other variables adjoint should just return their sampled/observed values. I don't know what has changed in funsor since this code was written but it seems like adjoint handles well those cases under the hood (see `# TODO this should really be handled entirely under the hood by adjoint` comment).

**Review by fehiepsi — APPROVED** [review]
